zilliqa/src/api/eth.rs

```diff
@@ -622,41 +750,42 @@ fn net_listening(_: Params, _: &Arc<Mutex<Node>>) -> Result<bool> {
 
 #[cfg(test)]
 mod tests {
-    use primitive_types::H160;
+    use primitive_types::U256;
 
     use crate::{
-        api::eth::{left_pad_arr, transaction_from_rlp},
+        api::eth::{left_pad_arr, parse_transaction},
+        crypto::Hash,
         state::Address,
+        transaction::{EthSignature, RecoveredTransaction, SignedTransaction, TxLegacy},
     };
 
     #[test]
     fn test_transaction_from_rlp() {
         // From https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md#example
         let transaction = hex::decode("f86c098504a817c800825208943535353535353535353535353535353535353535880de0b6b3a76400008025a028ef61340bd939bc2195fe537567866003e1a15d3c71ff63e1590620aa636276a067cbe9d8997f761aecb703304b3800ccf555c9f3dc64214b297fb1966a3b6d83").unwrap();
-        let signed_tx = transaction_from_rlp(&transaction, 1).unwrap();
-        let tx = &signed_tx.transaction;
-        assert_eq!(tx.nonce, 9);
-        assert_eq!(tx.gas_price, 20 * 10_u64.pow(9));
-        assert_eq!(tx.gas_limit, 21000u64);
-        assert_eq!(
-            tx.to_addr.unwrap(),
-            Address(
-                "0x3535353535353535353535353535353535353535"
-                    .parse::<H160>()
-                    .unwrap()
-            )
-        );
-        assert_eq!(tx.amount, 10u128.pow(18));
-        assert_eq!(tx.payload, Vec::<u8>::new());
-        assert_eq!(
-            signed_tx.from_addr,
-            Address(
-                "0x9d8A62f656a8d1615C1294fd71e9CFb3E4855A4F"
-                    .parse::<H160>()
-                    .unwrap()
-            )
-        );
-        assert!(signed_tx.verify().is_ok());
+        let signed_tx = parse_transaction(&transaction).unwrap();
+        let recovered_tx = signed_tx.recover_signer().unwrap();
+        let expected = RecoveredTransaction {
+            tx: SignedTransaction::Legacy {
+                tx: TxLegacy {
+                    chain_id: Some(1),
```

Comment:
am I missing something? Why does the legacy TX have a chain id, I thought the point was that it didn't?

Reply:
oh, is it that the TXs can handle pre and post EIP155?

Reply:
ah yes, I see my folly. Legacy refers to both. Although, in all of our tests aren't we only using legacy, anyway?

The coverage of this test seems limited - doesn't cover txs with no chain id for example? I have found in the past that you can get edge cases with padding where most but not all txs get verified, just a thought.

Reply:
Yep, what you said :)

I just including this unit test here, because the test case was included in EIP-155. Legacy transactions without a chain ID should be tested by `send_legacy_transaction_without_chain_id` in `zilliqa/tests/it/eth.rs`.